pkcs11module.c

```diff
@@ -149,6 +152,20 @@ PHP_METHOD(Module, __construct) {
     objval->initialised = true;
 }
 
+PHP_METHOD(Module, __destruct) {
+    pkcs11_object *objval = Z_PKCS11_P(ZEND_THIS);
+
+    CK_RV rv;
+
+    if (!objval->initialised)
+        return;
+
+    rv = objval->functionList->C_Finalize(NULL);
+    if ((rv != CKR_OK) && (rv != CKR_CRYPTOKI_NOT_INITIALIZED))
+        pkcs11_error(rv, "could not be C_Finalize()'d");
+
+    objval->initialised = false;
+}
```

Comment:
This is redundant since C_Finalize is already being called inpkcs11_shutdown()

Reply:
OK, I think got the point of pkcs11_shutdown(): first time I thought it was called only when  php-pkcs11 gets shutdown by PHP. I see it registers for the destructor too thanks to  DEFINE_MAGIC_FUNCS() / DECLARE_MAGIC_FUNCS().

 I missed it.

Then, what's about adding warning case pkcs11_error(rv, "could not be C_Finalize()'d"); within pkcs11_shutdown() ?